File: neetcode/go/206.reverse-linked-list.go
/*
 * @lc app=leetcode id=206 lang=golang
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	currNode := head
	var prevNode *ListNode
	var nextNode *ListNode

	for currNode != nil {
		nextNode = currNode.Next
		currNode.Next = prevNode
		prevNode = currNode
		currNode = nextNode
	}
	return prevNode
}

// @lc code=end

File: neetcode/python/206.reverse-linked-list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head

        if head.next is None:
            return head
 
        curr_node = head
        prev_node = None
        next_node = None

        while curr_node is not None:
            # update node.next
            next_node = curr_node.next
            curr_node.next = prev_node
            # update pointers
            prev_node = curr_node
            curr_node = next_node

        return prev_node

# @lc code=end

